testing/integration/enroll_replace_token_test.go

```diff
@@ -0,0 +1,118 @@
+	"context"
+	"testing"
+	"time"
+
+	"github.com/gofrs/uuid/v5"
+	"github.com/stretchr/testify/require"
+
+	"github.com/elastic/elastic-agent-libs/kibana"
+	atesting "github.com/elastic/elastic-agent/pkg/testing"
+	"github.com/elastic/elastic-agent/pkg/testing/define"
+	"github.com/elastic/elastic-agent/pkg/testing/tools"
+	"github.com/elastic/elastic-agent/pkg/testing/tools/check"
+	"github.com/elastic/elastic-agent/pkg/testing/tools/fleettools"
+)
+
+func TestEnrollReplaceToken(t *testing.T) {
+	info := define.Require(t, define.Requirements{
+		Group: Default,
+		Stack: &define.Stack{},
+		Sudo:  true,
+	})
+
+	ctx := context.Background()
+	fixture, err := define.NewFixtureFromLocalBuild(t, define.Version())
+	require.NoError(t, err)
+
+	fleetServerURL, err := fleettools.DefaultURL(ctx, info.KibanaClient)
+	require.NoError(t, err)
+
+	randId := uuid.Must(uuid.NewV4()).String()
+	policyReq := kibana.AgentPolicy{
+		Name:        "test-policy-" + randId,
+		Namespace:   "default",
+		Description: "Test policy " + randId,
+		MonitoringEnabled: []kibana.MonitoringEnabledOption{
+			kibana.MonitoringEnabledLogs,
+			kibana.MonitoringEnabledMetrics,
+		},
+	}
+	policy, err := info.KibanaClient.CreatePolicy(ctx, policyReq)
+	require.NoError(t, err)
+
+	enrollmentApiKey, err := tools.CreateEnrollmentToken(t, ctx, info.KibanaClient, policy.ID)
+	require.NoError(t, err)
+
+	// use a defined ID and replace token
+	agentID := uuid.Must(uuid.NewV4()).String()
+	replaceToken := uuid.Must(uuid.NewV4()).String()
+	installOpts := atesting.InstallOpts{
+		NonInteractive: true,
+		Force:          true,
+		Privileged:     false,
+		EnrollOpts: atesting.EnrollOpts{
+			ID:              agentID,
+			URL:             fleetServerURL,
+			EnrollmentToken: enrollmentApiKey.APIKey,
+			ReplaceToken:    replaceToken,
+		},
+	}
+
+	output, err := fixture.Install(ctx, &installOpts)
+	if err != nil {
+		t.Log(string(output))
+		t.Fatalf("failed installing the agent: %s", err)
+	}
+
+	t.Logf(">>> Enroll succeeded. Output: %s", output)
+
+	// Wait for Agent to be healthy
+	require.Eventually(
+		t,
+		check.FleetAgentStatusByAgentID(ctx, t, info.KibanaClient, agentID, "online"),
+		10*time.Minute,
+		10*time.Second,
+		"Elastic Agent status is not online",
+	)
+
+	// Force uninstall the Agent (skipping information Fleet)
+	output, err = fixture.Uninstall(ctx, &atesting.UninstallOpts{
+		Force:          true,
+		SkipFleetAudit: true,
+	})
+	if err != nil {
+		t.Log(string(output))
+		t.Fatalf("failed uninstalling the agent: %s", err)
+	}
+
+	// Using a clean fixture to ensure that no previous state is used
+	// re-enroll again and ensure that it all works again.
+	fixture, err = define.NewFixtureFromLocalBuild(t, define.Version())
+	require.NoError(t, err)
+
+	output, err = fixture.Install(ctx, &installOpts)
+	if err != nil {
+		t.Log(string(output))
+		t.Fatalf("failed installing the agent again: %s", err)
+	}
+
+	t.Logf(">>> Enroll succeeded again. Output: %s", output)
```

Comment:
you don't know if this enrollment is actually a different/newer one compared to the previous enrollment you did above. IMO, you need to extend `FleetAgentStatusByAgentID` to also expose and check enrolled_at ([api ref](https://www.elastic.co/docs/api/doc/serverless/operation/operation-get-fleet-agents-agentid#operation-get-fleet-agents-agentid-200-body-application-json-elastic-api-version-2023-10-31-item-enrolled_at)

Reply:
The `enrolled_at` doesn't change because in the eyes of Fleet. This Elastic Agent didn't become unenrolled and then enrolled so the `enrolled_at` is not reset, see here - https://github.com/elastic/fleet-server/pull/4417/files#diff-b3d2d3a49ef214fba08e3aa10772162cbb65c378e24ef6c4a0c5df65b051f71aR372

I should be able to set the headers on enrollment, to identify the difference.

Reply:
I don't think it should because it is not a new agent, it is just replacing an existing agent and becoming the same agent.

Reply:
I still have that in my head as something that should be updated or at least the comment in the code on the fleet-side makes me anticipate it 🙂 But I guess you know the intricacies of Fleet and Kibana UI better so if we did that we might even break something there. Let's skip it for now, and if required we might introduce even a new field to capture the replacements in the future?!